### PR TITLE
feat: update invokers to support swift 2

### PR DESCRIPTION
**Pull request by jaer-tsun**
**Reason for Change**:
updating invokers to support swift 2 contract changes

**Issue Fixed**:

**Requirements**:

- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [ ] relevant PR labels added

**Notes**:

### CODE REVIEW

File: cni/network/invoker_cns.go [hunk 1]
```diff
@@ -127,97 +132,167 @@ func (invoker *CNSIPAMInvoker) Add(addConfig IPAMAddConfig) (IPAMAddResult, erro
...
 		log.Logger.Info("Received info for pod",
-			zap.Any("ipv4info", info),
-			zap.Any("podInfo", podInfo))
-		ip, ncIPNet, err := net.ParseCIDR(info.podIPAddress + "/" + fmt.Sprint(info.ncSubnetPrefix))
-		if ip == nil {
-			return IPAMAddResult{}, errors.Wrap(err, "Unable to parse IP from response: "+info.podIPAddress+" with err %w")
-		}
+			zap.Any("ipinfo", info),
+			zap.Any("podInfo", podInfo),
+			zap.String("component", "cni-invoker-cns"))
```
Comment (tamilmani1989) [inline cni/network/invoker_cns.go:144]: no need to add component in every log.. check with paulyu on this

File: cni/network/invoker_cns.go [hunk 2]
```diff
@@ -127,97 +132,167 @@ func (invoker *CNSIPAMInvoker) Add(addConfig IPAMAddConfig) (IPAMAddResult, erro
...
-			if (invoker.ipamMode != util.V4Overlay) && (invoker.ipamMode != util.DualStackOverlay) && (invoker.ipamMode != util.Overlay) {
-				return IPAMAddResult{}, errors.Wrap(errInvalidArgs, "%w: Gateway address "+info.ncGatewayIPAddress+" from response is invalid")
+			macAddress, err := net.ParseMAC(info.macAddress)
+			if err != nil {
+				return IPAMAddResult{}, errors.Wrap(err, "Invalid mac address")
 			}
 
-			if net.ParseIP(info.podIPAddress).To4() != nil { //nolint:gocritic
-				ncgw, err = getOverlayGateway(ncIPNet)
+			isDefaultInterfaceSet = isDefaultInterfaceSet || info.isDefaultInterface
```
Comment (tamilmani1989) [inline cni/network/invoker_cns.go:158]: can you add a comment why you need this (or with local varaible)?

File: cni/network/invoker_cns.go [hunk 3]
```diff
@@ -127,97 +132,167 @@ func (invoker *CNSIPAMInvoker) Add(addConfig IPAMAddConfig) (IPAMAddResult, erro
...
+						},
+					},
+				},
+				addressType:        cns.Secondary,
+				macAddress:         macAddress,
+				isDefaultInterface: info.isDefaultInterface,
+			}
+
+			for _, route := range info.routes {
+				_, dst, err := net.ParseCIDR(route.IPAddress)
```
Comment (tamilmani1989) [inline cni/network/invoker_cns.go:181]: should we be using netip package instead of net?

Comment (tamilmani1989) [inline cni/network/invoker_cns.go:181]: ParseCIDR will convert to subnet format but not sure about /32 address. if intention is to add individual ip routes, will this work? For eg: if adding route for this ip 192.168.4.56/32, will parsecidr return as expected for this?

File: cni/network/invoker_cns.go [hunk 4]
```diff
@@ -127,97 +132,167 @@ func (invoker *CNSIPAMInvoker) Add(addConfig IPAMAddConfig) (IPAMAddResult, erro
...
+			}
 
-		ncgw := net.ParseIP(info.ncGatewayIPAddress)
-		if ncgw == nil {
-			// TODO: Remove v4overlay and dualstackoverlay options, after 'overlay' rolls out in AKS-RP
-			if (invoker.ipamMode != util.V4Overlay) && (invoker.ipamMode != util.DualStackOverlay) && (invoker.ipamMode != util.Overlay) {
-				return IPAMAddResult{}, errors.Wrap(errInvalidArgs, "%w: Gateway address "+info.ncGatewayIPAddress+" from response is invalid")
+			macAddress, err := net.ParseMAC(info.macAddress)
+			if err != nil {
+				return IPAMAddResult{}, errors.Wrap(err, "Invalid mac address")
```
Comment (tamilmani1989) [inline cni/network/invoker_cns.go:155]: why this is mandatory? if goalstate doesn't contain macaddress then cni should skip parsing it

File: cni/network/invoker_cns.go [hunk 5]
```diff
@@ -127,97 +132,167 @@ func (invoker *CNSIPAMInvoker) Add(addConfig IPAMAddConfig) (IPAMAddResult, erro
...
+					}
+
+					gw := net.ParseIP(route.GatewayIPAddress)
+					if gw == nil {
+						return IPAMAddResult{}, fmt.Errorf("unable to parse gateway %s: %w", route.GatewayIPAddress, routeErr)
+					}
+
+					addResult.defaultCniResult.ipResult.Routes = append(addResult.defaultCniResult.ipResult.Routes,
+						&cniTypes.Route{
+							Dst: *dst,
```
Comment (tamilmani1989) [inline cni/network/invoker_cns.go:269]: if code is same between primary and secondary, can we make it common?

File: cni/network/invoker_cns.go [hunk 6]
```diff
@@ -127,97 +132,167 @@ func (invoker *CNSIPAMInvoker) Add(addConfig IPAMAddConfig) (IPAMAddResult, erro
...
+					addResult.defaultCniResult.ipResult = &cniTypesCurr.Result{}
+				}
+
+				defaultRouteDstPrefix := network.Ipv4DefaultRouteDstPrefix
+				if ip.To4() == nil {
+					defaultRouteDstPrefix = network.Ipv6DefaultRouteDstPrefix
+					addResult.ipv6Enabled = true
+				}
+
+				addResult.defaultCniResult.ipResult.IPs = append(addResult.defaultCniResult.ipResult.IPs,
```
Comment (tamilmani1989) [inline cni/network/invoker_cns.go:246]: instead of referring `addResult.defaultCniResult.ipResult` in multiple places, what if we create a temp variable of type ipResult and then copy later to addResult.defaultCniResult

**Review by tamilmani1989 — CHANGES_REQUESTED** [review]

File: cni/network/invoker_cns.go (hunk 6 above)
Comment (jaer-tsun) [reply cni/network/invoker_cns.go:246]: we discussed a while back that this seems cleaner than having to loop through and find the default addressType every time

File: cni/network/invoker_cns.go (hunk 1 above)
Comment (jaer-tsun) [reply cni/network/invoker_cns.go:144]: removed after rebase

File: cni/network/invoker_cns.go (hunk 4 above)
Comment (jaer-tsun) [reply cni/network/invoker_cns.go:155]: I guess we can skip the err if we expect it mac to be valid, but this is just to be sure

File: cni/network/invoker_cns.go (hunk 2 above)
Comment (jaer-tsun) [reply cni/network/invoker_cns.go:158]: added comment at var init

File: cni/network/invoker_cns.go (hunk 6 above)
Comment (jaer-tsun) [reply cni/network/invoker_cns.go:246]: updated

File: cni/network/invoker_cns.go (hunk 3 above)
Comment (jaer-tsun) [reply cni/network/invoker_cns.go:181]: it works for /32 as well, dst will be 192.168.4.56/32 for your example

Comment (jaer-tsun) [reply cni/network/invoker_cns.go:181]: we'll need ParseCIDR since we expect input to be in CIDR notation

File: cni/network/invoker_cns.go (hunk 5 above)
Comment (jaer-tsun) [reply cni/network/invoker_cns.go:269]: using common func getRoutes now

File: cni/network/invoker_azure.go [hunk 7]
```diff
@@ -95,14 +96,17 @@ func (invoker *AzureIPAMInvoker) Add(addConfig IPAMAddConfig) (IPAMAddResult, er
 			nwCfg6.IPAM.Subnet = invoker.nwInfo.Subnets[1].Prefix.String()
 		}
 
-		addResult.ipv6Result, err = invoker.plugin.DelegateAdd(nwCfg6.IPAM.Type, &nwCfg6)
+		var ipv6Result *cniTypesCurr.Result
+		ipv6Result, err = invoker.plugin.DelegateAdd(nwCfg6.IPAM.Type, &nwCfg6)
 		if err != nil {
 			err = invoker.plugin.Errorf("Failed to allocate v6 pool: %v", err)
+		} else {
+			addResult.defaultCniResult.ipResult.IPs = append(addResult.defaultCniResult.ipResult.IPs, ipv6Result.IPs...)
```
Comment (tamilmani1989) [inline cni/network/invoker_azure.go:104]: can create a temp variable of this type `addResult.defaultCniResult.ipResult`  and later assign it to addResult.defaultcniresult since referring this in multiple places

File: cni/network/invoker_azure.go [hunk 8]
```diff
@@ -63,20 +61,23 @@ func (invoker *AzureIPAMInvoker) Add(addConfig IPAMAddConfig) (IPAMAddResult, er
...
+	addResult.defaultCniResult = CNIResult{ipResult: ipv4Result, addressType: cns.Default, isDefaultInterface: true}
+	if len(ipv4Result.IPs) > 0 {
+		addResult.hostSubnetPrefix = ipv4Result.IPs[0].Address
+	}
 
 	defer func() {
 		if err != nil {
-			if len(addResult.ipv4Result.IPs) > 0 {
-				if er := invoker.Delete(&addResult.ipv4Result.IPs[0].Address, addConfig.nwCfg, nil, addConfig.options); er != nil {
+			if len(addResult.defaultCniResult.ipResult.IPs) > 0 {
```
Comment (tamilmani1989) [inline cni/network/invoker_azure.go:79]: same comment as line 104

File: cni/network/invoker_cns.go [hunk 9]
```diff
@@ -120,6 +120,7 @@ func (invoker *CNSIPAMInvoker) Add(addConfig IPAMAddConfig) (IPAMAddResult, erro
 	}
 
 	addResult := IPAMAddResult{}
+	var isDefaultInterfaceSet bool // make sure isDefaultInterface is set accordingly for defaultCniResult
```
Comment (tamilmani1989) [inline cni/network/invoker_cns.go:123]: I couldn't understand anything from this comment. Please add detailed description how this flag will be set for different scenario

File: cni/network/invoker_cns.go (hunk 2 above)
Comment (tamilmani1989) [reply cni/network/invoker_cns.go:156]: its good to add comment here as well why its or with local variable

File: cni/network/invoker_cns.go [hunk 10]
```diff
@@ -130,97 +131,143 @@ func (invoker *CNSIPAMInvoker) Add(addConfig IPAMAddConfig) (IPAMAddResult, erro
...
-		}
 
-		ncgw := net.ParseIP(info.ncGatewayIPAddress)
-		if ncgw == nil {
-			// TODO: Remove v4overlay and dualstackoverlay options, after 'overlay' rolls out in AKS-RP
-			if (invoker.ipamMode != util.V4Overlay) && (invoker.ipamMode != util.DualStackOverlay) && (invoker.ipamMode != util.Overlay) {
-				return IPAMAddResult{}, errors.Wrap(errInvalidArgs, "%w: Gateway address "+info.ncGatewayIPAddress+" from response is invalid")
+		switch info.addressType {
+		case cns.Secondary:
+			ip, ipnet, err := response.PodIPInfo[i].PodIPConfig.GetIPNet()
```
Comment (tamilmani1989) [inline cni/network/invoker_cns.go:146]: i would suggest to move to different function for each case as function is growing big

File: cni/network/invoker_cns.go [hunk 11]
```diff
@@ -130,97 +131,143 @@ func (invoker *CNSIPAMInvoker) Add(addConfig IPAMAddConfig) (IPAMAddResult, erro
...
+					})
+				defaultCniResult.Routes = append(defaultCniResult.Routes,
+					&cniTypes.Route{
+						Dst: defaultRouteDstPrefix,
 						GW:  ncgw,
-					},
-				},
+					})
+
+				if err = getRoutes(defaultCniResult, info.routes); err != nil {
```
Comment (tamilmani1989) [inline cni/network/invoker_cns.go:243]: can `getRoutes` return routes and then assign it to cniresult? updating inside the function doesn't look correct

**Review by tamilmani1989 — CHANGES_REQUESTED** [review]

File: cni/network/invoker_azure.go (hunk 7 above)
Comment (jaer-tsun) [reply cni/network/invoker_azure.go:104]: just re-using ipv4Result and moved assignment to CNIResult at the end

File: cni/network/invoker_cns.go (hunk 9 above)
Comment (jaer-tsun) [reply cni/network/invoker_cns.go:123]: updated comment but we can also just default interface value in network.go if there are no secondary CNIResults

File: cni/network/invoker_cns.go (hunk 11 above)
Comment (jaer-tsun) [reply cni/network/invoker_cns.go:243]: yup, updated

File: cni/network/invoker_cns.go (hunk 2 above)
Comment (jaer-tsun) [reply cni/network/invoker_cns.go:160]: updated comment at init, it should make sense now

File: cni/network/invoker_cns.go (hunk 10 above)
Comment (tamilmani1989) [reply cni/network/invoker_cns.go:150]: not addressing this?

Comment (jaer-tsun) [reply cni/network/invoker_cns.go]: updated

File: cni/network/invoker_cns.go [hunk 12]
```diff
@@ -118,6 +122,8 @@ func (invoker *CNSIPAMInvoker) Add(addConfig IPAMAddConfig) (IPAMAddResult, erro
 	}
 
 	addResult := IPAMAddResult{}
+	// Default address type will be the default interface unless isDefaultInterface is true for a secondary address
```
Comment (neaggarwMS) [inline cni/network/invoker_cns.go:125]: This is a confusing comment. Ideally we should not be using isDefaultInterface to determine the routes

File: cns/NetworkContainerContract.go [hunk 13]
```diff
@@ -406,6 +412,14 @@ type PodIpInfo struct {
 	PodIPConfig                     IPSubnet
 	NetworkContainerPrimaryIPConfig IPConfiguration
 	HostPrimaryIPInfo               HostIPInfo
+	// NICType defines whether NIC is Default or Secondary
+	NICType string
+	// MacAddress of (secondary) interface
+	MacAddress string
+	// IsDefaultInterface is true if traffic should go through this interface by default
+	IsDefaultInterface bool
```
Comment (neaggarwMS) [inline cns/NetworkContainerContract.go:420]: SkipDefaultRoute bool (default: false)

Comment (neaggarwMS) [reply cns/NetworkContainerContract.go:420]: SkipDefaultRoutes

File: cns/NetworkContainerContract.go [hunk 14]
```diff
@@ -70,6 +70,12 @@ const (
 	Vxlan = "Vxlan"
 )
 
+// Address Types
+const (
+	Default   = "Default"
```
Comment (neaggarwMS) [inline cns/NetworkContainerContract.go:75]: Lets call it ManagementNic

Comment (neaggarwMS) [reply cns/NetworkContainerContract.go:75]: Update: call it InfraNic

File: cns/NetworkContainerContract.go [hunk 15]
```diff
@@ -70,6 +70,12 @@ const (
...
+	Secondary = "Secondary"
```
Comment (neaggarwMS) [inline cns/NetworkContainerContract.go:76]: calls SecondaryNic

File: cni/network/invoker_azure.go [hunk 16]
```diff
@@ -92,13 +92,18 @@ func (invoker *AzureIPAMInvoker) Add(addConfig IPAMAddConfig) (IPAMAddResult, er
...
 
-		addResult.ipv6Result, err = invoker.plugin.DelegateAdd(nwCfg6.IPAM.Type, &nwCfg6)
+		var ipv6Result *cniTypesCurr.Result
+		ipv6Result, err = invoker.plugin.DelegateAdd(nwCfg6.IPAM.Type, &nwCfg6)
 		if err != nil {
 			err = invoker.plugin.Errorf("Failed to allocate v6 pool: %v", err)
+		} else {
+			ipv4Result.IPs = append(ipv4Result.IPs, ipv6Result.IPs...)
+			ipv4Result.Routes = append(ipv4Result.Routes, ipv6Result.Routes...)
+			addResult.ipv6Enabled = true
```
Comment (tamilmani1989) [inline cni/network/invoker_azure.go:102]: this is confusing.. instead of reusing ipv4Result should define a `result` variable and then copy ipv4Result and ipv6Result to it

File: cni/network/invoker_azure_test.go [hunk 17]
```diff
@@ -22,36 +22,24 @@ type mockDelegatePlugin struct {
...
 	if pluginName == ipamV6 {
 		if d.errv6 != nil {
 			return nil, d.errv6
 		}
-		if d.resultsIPv6 == nil || d.resultsIPv6Index-1 > len(d.resultsIPv6) {
-			return nil, errors.New("no more ipv6 results in mock available") //nolint:goerr113
-		}
-		res := d.resultsIPv6[d.resultsIPv6Index]
-		d.resultsIPv6Index++
-		return res, nil
```
Comment (tamilmani1989) [inline cni/network/invoker_azure_test.go:43]: why did you remove this code?

File: cni/network/invoker_cns.go [hunk 18]
```diff
@@ -132,97 +133,44 @@ func (invoker *CNSIPAMInvoker) Add(addConfig IPAMAddConfig) (IPAMAddResult, erro
...
-					},
-				},
-				Routes: []*cniTypes.Route{
-					{
-						Dst: network.Ipv6DefaultRouteDstPrefix,
-						GW:  ncgw,
-					},
-				},
+		default:
+			// only count dualstack interface once
```
Comment (tamilmani1989) [inline cni/network/invoker_cns.go:157]: Is it typo? default interface or dualstack interface?

File: cni/network/invoker_cns.go [hunk 19]
```diff
@@ -132,97 +133,44 @@ func (invoker *CNSIPAMInvoker) Add(addConfig IPAMAddConfig) (IPAMAddResult, erro
...
+			overlayMode := (invoker.ipamMode == util.V4Overlay) || (invoker.ipamMode == util.DualStackOverlay) || (invoker.ipamMode == util.Overlay)
+			if err := configureDefaultAddResult(&info, &addConfig, &addResult, overlayMode); err != nil {
 				return IPAMAddResult{}, err
 			}
 		}
 	}
 
+	// Make sure default routes exist for 1 interface
+	if numInterfacesWithDefaultRoutes != 1 {
+		return IPAMAddResult{}, errors.New("Add result requires exactly one interface with default routes")
```
Comment (tamilmani1989) [inline cni/network/invoker_cns.go:171]: can we define a new error variable for this error?

File: cni/network/invoker_cns.go [hunk 20]
```diff
@@ -356,3 +304,158 @@ func (invoker *CNSIPAMInvoker) Delete(address *net.IPNet, nwCfg *cni.NetworkConf
...
+	// construct ipnet for result
+	resultIPnet := net.IPNet{
+		IP:   ip,
+		Mask: ncIPNet.Mask,
+	}
+
+	if ip := net.ParseIP(info.podIPAddress); ip != nil {
+		defaultInterfaceInfo := addResult.defaultInterfaceInfo.ipResult
+		if defaultInterfaceInfo == nil {
+			defaultInterfaceInfo = &cniTypesCurr.Result{}
```
Comment (tamilmani1989) [inline cni/network/invoker_cns.go:366]: in which case this will happen? looks to me over cautious.. if its not gonna happen, i would suggest to remove this check

File: cni/network/invoker_cns.go [hunk 21]
```diff
@@ -356,3 +304,158 @@ func (invoker *CNSIPAMInvoker) Delete(address *net.IPNet, nwCfg *cni.NetworkConf
...
+			return errors.Wrap(err, "No podIPAddress is found: %w")
+		}
+	}
+
+	// construct ipnet for result
+	resultIPnet := net.IPNet{
+		IP:   ip,
+		Mask: ncIPNet.Mask,
+	}
+
```
Comment (tamilmani1989) [inline cni/network/invoker_cns.go:362]: can you  move this to place where you append resultIPnet

File: cni/network/invoker_cns.go [hunk 22]
```diff
@@ -356,3 +304,158 @@ func (invoker *CNSIPAMInvoker) Delete(address *net.IPNet, nwCfg *cni.NetworkConf
...
+				},
+			},
+			Interfaces: []*cniTypesCurr.Interface{
+				{
+					Mac: info.macAddress,
+				},
+			},
+		},
+		nicType:           cns.DelegatedVMNIC,
+		macAddress:        macAddress,
```
Comment (tamilmani1989) [inline cni/network/invoker_cns.go:448]: Why do we need this if `cniTypesCurr.Interface` has `Mac` field?

File: cni/network/invoker_cns_test.go [hunk 23]
```diff
@@ -213,31 +212,118 @@ func TestCNSIPAMInvoker_Add_Overlay(t *testing.T) {
...
+						Dst: network.Ipv6DefaultRouteDstPrefix,
+						GW:  net.ParseIP("fe80::1234:5678:9abc"),
+					},
 				},
 			},
-			wantIpv6Result: &cniTypesCurr.Result{
+			wantErr: false,
+		},
+		{
+			name: "Test happy CNI add with multitenant result",
```
Comment (tamilmani1989) [inline cni/network/invoker_cns_test.go:240]: can you add test for failure case?

Comment (tamilmani1989) [inline cni/network/invoker_cns_test.go:240]: is test added for covering delegated nic scenario?

File: cni/network/invoker_mock.go [hunk 24]
```diff
@@ -53,9 +54,12 @@ func (invoker *MockIpamInvoker) Add(opt IPAMAddConfig) (ipamAddResult IPAMAddRes
 	ip := net.ParseIP(ipv4Str)
 	ipnet := net.IPNet{IP: ip, Mask: net.CIDRMask(subnetBits, ipv4Bits)}
 	gwIP := net.ParseIP("10.240.0.1")
-	ipConfig := &current.IPConfig{Address: ipnet, Gateway: gwIP}
-	ipamAddResult.ipv4Result = &current.Result{}
-	ipamAddResult.ipv4Result.IPs = append(ipamAddResult.ipv4Result.IPs, ipConfig)
+	ipamAddResult.defaultInterfaceInfo = InterfaceInfo{
```
Comment (tamilmani1989) [inline cni/network/invoker_mock.go:57]: is delegated nic covered?

**Review by tamilmani1989 — CHANGES_REQUESTED** [review]

File: cni/network/invoker_azure.go (hunk 16 above)
Comment (jaer-tsun) [reply cni/network/invoker_azure.go:102]: renamed to result and then appending v6 results like before

File: cni/network/invoker_cns.go (hunk 18 above)
Comment (jaer-tsun) [reply cni/network/invoker_cns.go:157]: since we get v4 and v6 pod ip info separately for dualstack, we just want to count this interface once

File: cni/network/invoker_cns.go (hunk 19 above)
Comment (jaer-tsun) [reply cni/network/invoker_cns.go:171]: added new errInvalidDefaultRouting

File: cni/network/invoker_cns.go (hunk 20 above)
Comment (jaer-tsun) [reply cni/network/invoker_cns.go:366]: this will be nil until we handle the default interface (call configureDefaultAddResult

File: cni/network/invoker_cns.go (hunk 21 above)
Comment (jaer-tsun) [reply cni/network/invoker_cns.go:362]: moved

File: cni/network/invoker_cns.go (hunk 22 above)
Comment (jaer-tsun) [reply cni/network/invoker_cns.go:448]: removed

File: cni/network/invoker_cns_test.go (hunk 23 above)
Comment (jaer-tsun) [reply cni/network/invoker_cns_test.go:240]: yup

File: cni/network/invoker_mock.go (hunk 24 above)
Comment (jaer-tsun) [reply cni/network/invoker_mock.go:63]: added

File: cni/network/invoker_cns_test.go (hunk 23 above)
Comment (jaer-tsun) [reply cni/network/invoker_cns_test.go:243]: added

File: cni/network/invoker_azure.go [hunk 25]
```diff
@@ -92,13 +92,18 @@ func (invoker *AzureIPAMInvoker) Add(addConfig IPAMAddConfig) (IPAMAddResult, er
...
 
-		addResult.ipv6Result, err = invoker.plugin.DelegateAdd(nwCfg6.IPAM.Type, &nwCfg6)
+		var ipv6Result *cniTypesCurr.Result
+		ipv6Result, err = invoker.plugin.DelegateAdd(nwCfg6.IPAM.Type, &nwCfg6)
 		if err != nil {
 			err = invoker.plugin.Errorf("Failed to allocate v6 pool: %v", err)
+		} else {
+			result.IPs = append(result.IPs, ipv6Result.IPs...)
+			result.Routes = append(result.Routes, ipv6Result.Routes...)
+			addResult.ipv6Enabled = true
```
Comment (behzad-mir) [inline cni/network/invoker_azure.go:102]: Shouldn't we first check if Ipv6 result is empty or not before setting addResult.ipv6Enabled = true ?

File: cni/network/invoker_cns.go [hunk 26]
```diff
@@ -132,97 +135,44 @@ func (invoker *CNSIPAMInvoker) Add(addConfig IPAMAddConfig) (IPAMAddResult, erro
...
-			addConfig.options[network.SNATIPKey] = info.ncPrimaryIP
+			nicType:            response.PodIPInfo[i].NICType,
+			macAddress:         response.PodIPInfo[i].MacAddress,
+			skipDefaultRoutes:  response.PodIPInfo[i].SkipDefaultRoutes,
+			routes:             response.PodIPInfo[i].Routes,
 		}
 
 		logger.Info("Received info for pod",
-			zap.Any("ipv4info", info),
+			zap.Any("ipinfo", info),
```
Comment (behzad-mir) [inline cni/network/invoker_cns.go:145]: neat : ipInfo?

File: cni/network/invoker_cns.go [hunk 27]
```diff
@@ -132,97 +135,44 @@ func (invoker *CNSIPAMInvoker) Add(addConfig IPAMAddConfig) (IPAMAddResult, erro
...
-		if ip == nil {
-			return IPAMAddResult{}, errors.Wrap(err, "Unable to parse IP from response: "+info.podIPAddress+" with err %w")
-		}
 
-		ncgw := net.ParseIP(info.ncGatewayIPAddress)
-		if ncgw == nil {
-			// TODO: Remove v4overlay and dualstackoverlay options, after 'overlay' rolls out in AKS-RP
-			if (invoker.ipamMode != util.V4Overlay) && (invoker.ipamMode != util.DualStackOverlay) && (invoker.ipamMode != util.Overlay) {
-				return IPAMAddResult{}, errors.Wrap(errInvalidArgs, "%w: Gateway address "+info.ncGatewayIPAddress+" from response is invalid")
+		//nolint:exhaustive // ignore exhaustive types check
```
Comment (behzad-mir) [inline cni/network/invoker_cns.go:148]: neat: add an space: 
 // nolint...

File: cni/network/invoker_cns.go [hunk 28]
```diff
@@ -132,97 +135,44 @@ func (invoker *CNSIPAMInvoker) Add(addConfig IPAMAddConfig) (IPAMAddResult, erro
...
-			if err := setHostOptions(ncIPNet, addConfig.options, &info); err != nil {
+			overlayMode := (invoker.ipamMode == util.V4Overlay) || (invoker.ipamMode == util.DualStackOverlay) || (invoker.ipamMode == util.Overlay)
+			if err := configureDefaultAddResult(&info, &addConfig, &addResult, overlayMode); err != nil {
 				return IPAMAddResult{}, err
 			}
 		}
 	}
 
+	// Make sure default routes exist for 1 interface
+	if numInterfacesWithDefaultRoutes != 1 {
```
Comment (behzad-mir) [inline cni/network/invoker_cns.go:172]: neat: Better to use the magic numbers as const

**Review by behzad-mir — CHANGES_REQUESTED** [review]

File: cni/network/invoker_mock.go [hunk 29]
```diff
@@ -70,12 +80,24 @@ func (invoker *MockIpamInvoker) Add(opt IPAMAddConfig) (ipamAddResult IPAMAddRes
...
+		ipamAddResult.defaultInterfaceInfo.ipResult.IPs = append(ipamAddResult.defaultInterfaceInfo.ipResult.IPs, &current.IPConfig{Address: ipnet, Gateway: gwIP})
 		invoker.ipMap[ipnet.String()] = true
 	}
 
+	if invoker.delegatedVMNIC {
+		if invoker.delegatedVMNICFail {
+			return IPAMAddResult{}, errDelegatedVMNIC
+		}
+
+		ipStr := "20.20.20.20/32"
```
Comment (behzad-mir) [inline cni/network/invoker_mock.go:92]: plz make this a const

File: cni/network/invoker_cns.go (hunk 26 above)
Comment (jaer-tsun) [reply cni/network/invoker_cns.go:145]: updated

File: cni/network/invoker_cns.go (hunk 28 above)
Comment (jaer-tsun) [reply cni/network/invoker_cns.go:172]: sure, updated

File: cni/network/invoker_cns.go (hunk 27 above)
Comment (jaer-tsun) [reply cni/network/invoker_cns.go:152]: this seems standard for nolint

File: cni/network/invoker_cns.go (hunk 18 above)
Comment (tamilmani1989) [reply cni/network/invoker_cns.go:163]: ok then why are we not doing similar check for secondary interface?

File: cni/network/invoker_cns.go (hunk 20 above)
Comment (tamilmani1989) [reply cni/network/invoker_cns.go:371]: can `configureDefaultAddResult` be called with `addResult.defaultInterfaceInfo.ipResult` as nil?

File: cni/network/invoker_cns.go (hunk 22 above)
Comment (tamilmani1989) [reply cni/network/invoker_cns.go:451]: is this removed? Why am i still seeing this code then? my ask is why can't we reuse `cniTypesCurr.Interface` and avoid adding `macAddress` field separately

File: cni/network/network_test.go [hunk 30]
```diff
@@ -1103,3 +1103,75 @@ func TestGetPodSubnetNatInfo(t *testing.T) {
...
+				StdinData:   localNwCfg.Serialize(),
+				ContainerID: "test-container",
+				Netns:       "test-container",
+				Args:        fmt.Sprintf("K8S_POD_NAME=%v;K8S_POD_NAMESPACE=%v", "test-pod", "test-pod-ns"),
+				IfName:      eth0IfName,
+			},
+			wantErr: true,
+		},
+	}
+
```
Comment (tamilmani1989) [inline cni/network/network_test.go:1162]: I like to have testcase where defaultNic succeeded but secondary failed and cni should tear down dfaultNic setup and cleanup all allocated resources and vice-versa(secondary succeeded but default failed)

**Review by tamilmani1989 — CHANGES_REQUESTED** [review]

File: cni/network/invoker_cns.go (hunk 18 above)
Comment (jaer-tsun) [reply cni/network/invoker_cns.go:163]: we're not handling v6 for secondary interface at the moment

not sure if we want to change the current behaviour where the v4 and v6 ip info are sent separately for the same interface

File: cni/network/invoker_cns.go (hunk 20 above)
Comment (jaer-tsun) [reply cni/network/invoker_cns.go:371]: yes, the first call to configureDefaultAddResult will have nil addResult.defaultInterfaceInfo.ipResult

File: cni/network/invoker_azure_test.go [hunk 31]
```diff
@@ -82,8 +82,8 @@ func getCIDRNotationForAddress(ipaddresswithcidr string) *net.IPNet {
 	return ipnet
 }
 
-func getResult(ip string) []*cniTypesCurr.Result {
-	res := []*cniTypesCurr.Result{
+func getSingleResult(ip string) []*cniTypesCurr.Result {
```
[the remark this answers is not in the document]
Comment (jaer-tsun) [reply cni/network/invoker_azure_test.go:86]: updated

File: cni/network/network_test.go (hunk 30 above)
Comment (jaer-tsun) [reply cni/network/network_test.go:1162]: added via mock endpoint client, will add via transparent and secondary endpoint client in the future

File: cni/network/invoker_cns.go (hunk 22 above)
Comment (jaer-tsun) [reply cni/network/invoker_cns.go:452]: we removed the cniTypesCurr.Interface and kept macAddress field since the one in cniTypesCurr.Interface is a string while we need the parsed macAddress for endpointInfo

File: cni/network/network_test.go (hunk 30 above)
Comment (tamilmani1989) [reply cni/network/network_test.go:1162]: can you create work item for it? This test case is must for cni release

File: cni/network/invoker_cns.go [hunk 32]
```diff
@@ -21,11 +21,17 @@ import (
 	"go.uber.org/zap"
 )
 
+const (
+	ExpectedNumInterfacesWithDefaultRoutes = 1
```
Comment (tamilmani1989) [inline cni/network/invoker_cns.go:25]: why this is exported variable. can it be  `expectedNumInterfacesWithDefaultRoutes`?

File: cni/network/invoker_cns.go (hunk 20 above)
Comment (tamilmani1989) [reply cni/network/invoker_cns.go:372]: why so? If thats the case, can this check be even before this function called?

File: cni/network/invoker_cns.go (hunk 22 above)
Comment (tamilmani1989) [reply cni/network/invoker_cns.go:452]: but why cant we reuse `cniTypesCurr.Result.Interface` structure? then no need for new field `macAddress`

Comment (jaer-tsun) [reply cni/network/invoker_cns.go:451]: macAddress in cniTypesCurr.Result.Interface is string but we need net.HardwareAddr for EndpointInfo, just passing the net.HardwareAddr as new field since we already parsed mac here

File: cni/network/invoker_cns.go [hunk 33]
```diff
@@ -356,3 +314,151 @@ func (invoker *CNSIPAMInvoker) Delete(address *net.IPNet, nwCfg *cni.NetworkConf
...
+	}
+
+	macAddress, err := net.ParseMAC(info.macAddress)
+	if err != nil {
+		return errors.Wrap(err, "Invalid mac address")
+	}
+
+	result := InterfaceInfo{
+		ipResult: &cniTypesCurr.Result{
+			IPs: []*cniTypesCurr.IPConfig{
```
Comment (tamilmani1989) [inline cni/network/invoker_cns.go:441]: didn't we agree changing this to []IPs, []Routes instead of using Result type

**Review by tamilmani1989 — CHANGES_REQUESTED** [review]

File: cni/network/invoker_cns.go (hunk 33 above)
Comment (jaer-tsun) [reply cni/network/invoker_cns.go:441]: []Interfaces is still being used in parts of code, if we really need this refactor we should do in separate PR...

Comment (jaer-tsun) [reply cni/network/invoker_cns.go:441]: discussed offline to not take dependency on cniTypesCurr.Result and will split in separate PR

**Review by tamilmani1989 — APPROVED** [review]
